kernel/nativecode.ml

```diff
@@ -1653,12 +1689,13 @@ let pp_mllam fmt l =
     match l with
     | MLlocal ln -> Format.fprintf fmt "@[%a@]" pp_lname ln
     | MLglobal g -> Format.fprintf fmt "@[%a@]" pp_gname g
-    | MLprimitive p -> Format.fprintf fmt "@[%a@]" pp_primitive p
+    | MLprimitive (p, args) ->
+      Format.fprintf fmt "@[%a@ %a@]" pp_primitive p (pp_args true) args
     | MLlam(ids,body) ->
         Format.fprintf fmt "@[(fun%a@ ->@\n %a)@]"
           pp_ldecls ids pp_mllam body
     | MLletrec(defs, body) ->
-        Format.fprintf fmt "@[%a@ in@\n%a@]" pp_letrec defs
+        Format.fprintf fmt "@[(%a@ in@\n%a)@]" pp_letrec defs
```

Comment:
I realized that I kind of lied, this line changes the literal output of the native compiler. It's only a matter of parentheses, so I hope it doesn't matter.

Reply:
As I explained in the other pull request, with some versions of OCaml, using `Obj.magic` disables inlining. On such trivial functions, we cannot afford to have the compiler use closures. It would have a tremendous impact on performance.

Reply:
Can we not sprinkle `[@@inline]` where needed? It is just flambda that does this right?

Reply:
By the way, I am quite worried by the behavior of the compiler here. I tried the following snippet to see how the compiler behaved:
```ocaml
type t = t -> t
type 'a ty_arity = unit
let ar_0 : t ty_arity = ()
let ar_S (_ : 'a ty_arity) : (t -> 'a) ty_arity = ()
let ar_1 = ar_S ar_0
let ar_2 = ar_S ar_1
let app (_ : 'a ty_arity) (f : t) : 'a = Obj.magic f

let add accu x y = app ar_2 accu x y
```

The compiler complains that `x` and `y` are not used by the function `add` (and this seems reflected in the generated code). This is obviously wrong. Did I mess something with my testcase?

Reply:
Your remark is unrelated to this particular comment, which is about parentheses of letrec blocks. I don't know whether it has any effect on e.g. commutative cuts.

Regarding your particular warning issue, I can't reproduce and the CI is fine with the current code. Why would the generated code behave weirdly?

Reply:
Perhaps it is a bug in OCaml 4.13.1, but the fact is that the compiler warns that it is not passing `x` and `y` as arguments to the accumulator.

> Your remark is unrelated to this particular comment,

I know. It was supposed to be attached to your comment saying that the new functions are identity functions, which the compiler says that they are not. I don't know how to move a remark around.

Reply:
Almost of the instances where an explicit cast was added are functions that are already assuming that their argument is a closure. Except if we expect the compiler to inline these functions (which it probably does not) then this is zero-cost. Furthermore the would-be closure is always an accumulator, so I don't even think we generate "real" OCaml closures, they are always generated by our homegrown set-tag-to-0 hack.